spec/lib/site_cache_spec.rb

```diff
@@ -0,0 +1,261 @@
+
+    public_send(operation)
+    @cache.operations.last
+  end
+
+  def generate_operation
+    operation =
+      path.choose(
+        :insert_new,
+        :insert_existing_site_id,
+        :insert_existing_key,
+        :lookup_existing,
+        :getset_existing,
+        :delete_existing,
+        :clear,
+        :clear_existing_site,
+      )
+
+    public_send(operation)
+    @cache.operations.last
+  end
+
+  def generate_operations(upto:)
+    CacheOperations.new(path.choose(*(1..upto)).times.map { generate_operation })
+  end
+
+  def insert_new
+    site_id = Object.new
+    key = Object.new
+    keys[site_id] << key
+    value = Object.new
+
+    if path.choose(true, false)
+      cache.set(site_id, key, value)
+    else
+      cache.getset(site_id, key) { value }
+    end
+    [site_id, key, value]
+  end
+
+  def insert_existing_site_id
+    site_id = path.choose(*keys.keys)
+    key = Object.new
+    keys[site_id] << key
+    value = Object.new
+
+    if path.choose(true, false)
+      cache.set(site_id, key, value)
+    else
+      cache.getset(site_id, key) { value }
+    end
+  end
+
+  def insert_existing_key
+    site_id = path.choose(*keys.keys)
+    key = path.choose(*keys[site_id])
+    value = Object.new
+    cache.set(site_id, key, value)
+  end
+
+  def lookup_existing(site_id = nil)
+    site_id ||= path.choose(*keys.keys)
+    key = path.choose(*keys[site_id])
+    cache.lookup(site_id, key)
+  end
+
+  def getset_existing
+    site_id = path.choose(*keys.keys)
+    key = path.choose(*keys[site_id])
+    value = Object.new
+    cache.getset(site_id, key) { value }
+  end
+
+  def delete_existing(site_id = nil)
+    site_id ||= path.choose(*keys.keys)
+    key = path.choose(*keys[site_id])
+    cache.delete(site_id, key)
+  end
+
+  def clear
+    cache.clear
+  end
+
+  def clear_site(site_id)
+    cache.clear_site(site_id)
+  end
+
+  def clear_existing_site
+    site_id = path.choose(*keys.keys)
+    cache.clear_site(site_id)
+  end
+end
+
+class CacheOperations
+  def initialize(operations)
+    @operations = operations
+  end
+
+  def values_inserted_at(site_id, key)
+    @operations.filter_map do |op|
+      case op.method_name
+      when :set
+        op_site_id, op_key, op_value = op.args
+        op_value if site_id == op_site_id && key == op_key
+      when :getset
+        op_site_id, op_key = op.args
+        op_value = op.blk.call
+        op_value if site_id == op_site_id && key == op_key && op_value == op.result
+      end
+    end
+  end
+end
+
+RSpec.describe SiteCache do
+  it "produces values from lookups and deletes that were previously inserted" do
+    CachePathGenerator.complete(3, 2) do |g|
+      operations = g.generate_operations(upto: 4)
+      last_op = g.generate_lookup
+
+      site_id, key = last_op.args
+      result = last_op.result
+
+      expect(operations.values_inserted_at(site_id, key)).to include(result) if result
+    end
+  end
```

Comment:
Just taking this test as an example, I feel like it isn't very readable making it hard to understand what exactly we are testing for.

Reply:
If you have specific suggestions, I'd be interested in hearing them.